proptools/compare.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proptools
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// deepCompare is like reflect.deepEqual but it gives an explanation of the first difference that it finds
+func DeepCompare(a NamedItem, b NamedItem) (comparison Comparison) {
+	return deepCompareGeneric(toReflection(a), toReflection(b), 0)
+}
+
+type NamedItem struct {
+	Name string
+	Item interface{}
+}
+
+type Comparison struct {
```

Comment:
When is that (two values preferable to a struct) the case? Is it when the two values have fundamentally different meanings? It seems to me that a Comparison is itself a concept. If we wanted to instead return an entire list of differences, it'd be convenient to not have to update the signatures of each method based on the implementation of a Comparison.